hare/messagevalidation.go

```diff
@@ -103,30 +103,48 @@ func (validator *syntaxContextValidator) ContextuallyValidateMessage(m *Msg, cur
 		return false, errors.New("nil inner message")
 	}
 
-	// PreRound & Notify are always contextually valid
+	// first validate pre-round and notify
 	switch m.InnerMsg.Type {
-	case PreRound:
+	case Pre:
 		return true, nil
-	case Proposal:
-		if currentK == Round1 {
-			return false, errEarlyMsg
-		}
-		if currentK == Round2 || currentK == Round3 {
-			return true, nil // process proposals for
-		}
-		return false, nil
 	case Notify:
 		return true, nil
 	}
 
-	// Status & Commit Messages should match the expected K
-	if currentK == m.InnerMsg.K { // arrived on time
-		return true, nil
+	// the message must match the current iteration unless it is a notify message
+	currentIteration := currentK / 4
+	msgIteration := m.InnerMsg.K / 4
+	if currentIteration != msgIteration {
+		return false, nil
 	}
 
-	// matches the next round, early message
-	if currentK+1 == m.InnerMsg.K {
-		return false, errEarlyMsg
+	// Pre & Notify are always contextually valid
+	currentRound := currentK % 4
+	switch m.InnerMsg.Type {
+	case Proposal:
+		if currentRound == StatusRound {
+			return false, errEarlyMsg
+		}
+		if currentRound == ProposalRound || currentRound == CommitRound { // may be a late proposal
+			return true, nil
+		}
+		return false, nil
+	case Status:
+		if currentRound == PreRound {
+			return false, errEarlyMsg
+		}
+		if currentRound == StatusRound {
+			return true, nil
+		}
+		return false, nil
+	case Commit:
+		if currentRound == ProposalRound {
```

Comment:
and if it was received during Status round, don't you want to report `errEarlyMsg`?

Reply:
No. An early message is a message received one round prior to its expected round.
In the above case I return errInvalidRound